Updates based on recent changes and discussion on RFC PR #262 (#5)
annarev/community: rfcs/20200612-stream-executor-c-api/C_API_versioning_strategy.md

### SE_MAJOR
* Potentially backwards incompatible changes.
* If a change is backwards incompatible, it requires an RFC because it will
  break all current plug-ins. This should be rare.
* An `SE_MAJOR` update should be planned in a way that bundles as many pending
  backwards incompatible changes together as possible to avoid breaking plug-ins
  multiple times.
* There will be an announcement giving a grace period before the update happens.

### SE_MINOR
* Backwards compatible changes.
  * Adding a new variable, struct, method, enumeration, etc.
  * Trivial deprecation of a variable, etc. by setting it to a no-op values,
    e.g., 0 or `NULL`.

### SE_PATCH
* Backwards compatible bug fixes.

## Conventions
* Once a member is added to a struct, it cannot be removed, reordered, renamed,
  or repurposed (i.e., assigned a different functionality).
* "Renaming" a member is equivalent to adding a new member with a new name and
  eventually deprecating the member with the old name.
* Fields that cannot be 0 or `NULL` can be deprecated in a backwards compatible
  manner by zero-initialization. 
  * Plug-ins must perform input validation on these fields for 0 and `NULL`

>     before consuming them.

Reviewer: Should we have another bullet point for fields set by plugin? This case would be more common since most structs we have are populated by plug-in and not TensorFlow. In this case it should be something like:

TensorFlow will validate if any field which should be is deprecated by setting it to `0` or `NULL` is set to a different value and print a warning.

Author: You're right. How does "Both core TensorFlow and plug-ins" sounds? Both sides should validate.

Reviewer: I feel like its a bit different since TF is responsible for deprecation in both cases.

* If field is set by TF, then plug-in needs to check if it is 0 before using.
* If field is set by plug-in and TF deprecates it, then TF would never use it anyways. So "validate for 0 before consuming" is a noop. Instead, makes more sense if TF checks it is "non-zero" and print warning.

Author: True. Added what you said. :)

      * Plug-ins know the fields are deprecated when they find 0 or `NULL` in
        these fields.
  * Such fields must be explicitly marked by comments, to ensure all plug-ins

>     have consistent behavior (e.g., non of the plug-ins is using 0 or `NULL` as

Reviewer: nit. non --> none

Author: Done.

  * Trivial deprecation of a variable, etc. by setting it to a no-op values,
    e.g., 0 or `NULL`.

### SE_PATCH
* Backwards compatible bug fixes.

## Conventions
* Once a member is added to a struct, it cannot be removed, reordered, renamed,
  or repurposed (i.e., assigned a different functionality).
* "Renaming" a member is equivalent to adding a new member with a new name and
  eventually deprecating the member with the old name.
* Fields that cannot be 0 or `NULL` can be deprecated in a backwards compatible
  manner by zero-initialization. 
  * If the field is set by core TensorFlow, plug-ins must perform input validation
    on these fields for 0 and `NULL` before accessing them.
      * Plug-ins know the fields are deprecated when they find 0 or `NULL` in
        these fields.
  * If the field is set by plug-in, TF can check if the field is non-zero (or not
    `NULL`) and print a warning if so.
  * Such fields must be explicitly marked by comments, to ensure all plug-ins
    have consistent behavior (e.g., none of the plug-ins is using 0 or `NULL` as
    a special case). See `// 0 is no-op` and `// NULL is no-op` in the 

>     [By value inspection](#By-value-inspection) section for example.

Reviewer 2: ```suggestion
    [By value inspection](#by-value-inspection) section for example.
```

initialize_plugin_fn(&params, &status);
if(!tensorflow::StatusFromTF_Status(status).ok()) {
  // Output error and skip this plug-in.
}
```

Plug-in checks the `SE_MAJOR` version numbers and outputs error if they don't
match:
```c++
void SE_InitPlugin(SE_PlatformRegistrationParams* params,
                         TF_Status* status) {
  if (params->struct_size == 0) {
    // *status = ...
    LOG(ERROR) << "Invalid argument.";
    return;
  }
  if (SE_MAJOR != params->major) {
    // *status = ...
    LOG(ERROR) << "Unsupported major version. Given: " << params->major
               << " Expected: " << SE_MAJOR;
    return;
  }
  ...
}
```

### By Value Inspection
Deprecation of an attribute can sometimes be done in a backwards compatible
manner by leaving the attribute zero initialized.

* The plugin performs input validation on each field for `NULL` or 0 value
  before consuming it, preventing it from entering a bad state.
* If deprecation by zero-initialization is not possible (e.g., because default
  value of zero may be a valid input), then the change is API incompatible;
  TensorFlow has to bump the major version when the attribute is deprecated.

For example,

```c++

> struct example {

Reviewer 2: ```suggestion
struct Example {
```

>   int cannot_be_zero;  // 0 is no-op.
>   void* cannot_be_null;  // NULL is no-op.
>   int can_be_zero;
>   void* can_be_null;
>   int optional_zero_default;  // Optional. 0 by default.
>   void* optional_null_default;  // Optional. NULL by default.

Reviewer 2: ```suggestion
  int32_t cannot_be_zero;         // 0 is no-op.
  void* cannot_be_null;           // NULL is no-op.
  int32_t can_be_zero;
  void* can_be_null;
  int32_t optional_zero_default;  // Optional. 0 by default.
  void* optional_null_default;    // Optional. NULL by default.
```

Author: This is great. Thank you!
